File: src/center-service/app/controllers/management_request_controller.rb
#
# The MIT License
# Copyright (c) 2015 Estonian Information System Authority (RIA), Population Register Centre (VRK)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

require 'java'
require 'thread'

java_import Java::ee.ria.xroad.common.request.ClientRegRequestStatusWrapper
java_import Java::ee.ria.xroad.common.request.ManagementRequestHandler
java_import Java::ee.ria.xroad.common.request.ManagementRequestParser
java_import Java::ee.ria.xroad.common.request.ManagementRequestUtil
java_import Java::ee.ria.xroad.common.request.ManagementRequests
java_import Java::ee.ria.xroad.common.message.SoapUtils
java_import Java::ee.ria.xroad.common.message.SoapFault
java_import Java::ee.ria.xroad.common.ErrorCodes
java_import Java::ee.ria.xroad.common.CodedException
java_import Java::ee.ria.xroad.common.validation.SpringFirewallValidationRules

class ManagementRequestController < ApplicationController

    def create
        begin
            response.content_type = "text/xml"

            @xroad_instance = SystemParameter.instance_identifier
            raise "X-Road instance must exist!" if @xroad_instance.blank?

            @client_reg_request_status_wrapper = ClientRegRequestStatusWrapper.new

            @request_soap = ManagementRequestHandler.readRequest(
                request.headers["CONTENT_TYPE"],
                StringIO.new(request.raw_post).to_inputstream,
                @client_reg_request_status_wrapper)

            id = handle_request
            logger.debug("Created request id: #{id}")

            # Simply convert request message to response message
            response_soap = ManagementRequestUtil.toResponse(@request_soap, id)

            render :text => response_soap.getXml()
        rescue Java::java.lang.Exception => e
            handle_error(ErrorCodes.translateException(e))
        rescue Exception => e
            handle_error(CodedException.new(ErrorCodes::X_INTERNAL_ERROR, e.message))
            logger.error("Internal error: #{e.message}\n#{e.backtrace.join("\n\t")}")
        end
    end

    private

    def handle_request
        raise "Unknown service"
    end

    def handle_error(ex)
        render :text => SoapFault.createFaultXml(ex)
    end

    def security_server_id(id_type)
        SecurityServerId.from_parts(id_type.getXRoadInstance(),
                                    id_type.getMemberClass(), id_type.getMemberCode(),
                                    id_type.getServerCode())
    end

    def client_id(id_type)
        ClientId.from_parts(id_type.getXRoadInstance(), id_type.getMemberClass(),
                            id_type.getMemberCode(), id_type.getSubsystemCode())
    end

    def member_id(id_type)
        ClientId.from_parts(id_type.getXRoadInstance(), id_type.getMemberClass(),
                            id_type.getMemberCode())
    end

    def verify_owner(security_server)
        sender = client_id(@request_soap.getClient())
        verify_xroad_instance(sender)

        if not security_server.matches_client_id(sender)
            raise I18n.t("request.server_id_not_match_owner",
                         :security_server => security_server.to_s,
                         :sec_serv_owner => sender.to_s)
        end
    end

    def check_security_server_identifiers(ss)
      if ss.blank?
        check_identifier(ss.xroad_instance)
        check_identifier(ss.member_class)
        check_identifier(ss.member_code)
        check_identifier(ss.server_code)
      end
    end

    def check_client_identifiers(ss)
      if ss.blank?
        check_identifier(ss.xroad_instance)
        check_identifier(ss.member_class)
        check_identifier(ss.member_code)
        check_identifier(ss.subsystem_code)
      end
    end

    def check_identifier(id)
      unless id.blank?
        if SpringFirewallValidationRules::containsPercent(id) ||
          SpringFirewallValidationRules::containsSemicolon(id) ||
          SpringFirewallValidationRules::containsForwardslash(id) ||
          SpringFirewallValidationRules::containsBackslash(id) ||
          !SpringFirewallValidationRules::isNormalized(id)
          raise I18n.t("request.invalid_identifier", :id => id)
        end
      end
    end

    # xroad_id may be either ClientId or ServerId.
    def verify_xroad_instance(xroad_id)
        logger.debug("Instance verification: #{xroad_id}")

        unless @xroad_instance.eql?(xroad_id.xroad_instance)
            raise t("request.incorrect_instance")
        end
    end
end

File: src/central-server/management-service/infra-api-soap/src/main/java/org/niis/xroad/cs/management/api/ManagementRequestController.java
/**
 * The MIT License
 * <p>
 * Copyright (c) 2019- Nordic Institute for Interoperability Solutions (NIIS)
 * Copyright (c) 2018 Estonian Information System Authority (RIA),
 * Nordic Institute for Interoperability Solutions (NIIS), Population Register Centre (VRK)
 * Copyright (c) 2015-2017 Estonian Information System Authority (RIA), Population Register Centre (VRK)
 * <p>
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * <p>
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * <p>
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.niis.xroad.cs.management.api;

import ee.ria.xroad.common.CodedException;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.niis.xroad.common.managementrequest.ManagementRequestSoapExecutor;
import org.niis.xroad.common.managementrequest.model.ManagementRequestType;
import org.niis.xroad.cs.management.core.api.ManagementRequestService;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.RestController;

import java.io.InputStream;

import static ee.ria.xroad.common.ErrorCodes.X_INVALID_REQUEST;

@RestController
@Slf4j
@RequiredArgsConstructor
public class ManagementRequestController {
    private final ManagementRequestService managementRequestService;

    @ResponseBody
    @PostMapping(path = "/managementservice/manage",
            produces = {MediaType.TEXT_XML_VALUE},
            consumes = {MediaType.MULTIPART_RELATED_VALUE, MediaType.TEXT_XML_VALUE + ";charset=UTF-8"})
    public ResponseEntity<String> addManagementRequest(@RequestHeader(HttpHeaders.CONTENT_TYPE) String contentType, InputStream body) {
        return ManagementRequestSoapExecutor.process(contentType, body,
                result -> {
                    Integer requestId;
                    if (ManagementRequestType.AUTH_CERT_DELETION_REQUEST == result.getRequestType()) {
                        var authCertDeletionRequest = result.getAuthCertDeletionRequest()
                                .orElseThrow(() -> new CodedException(X_INVALID_REQUEST, "AuthCertDeletionRequest is missing"));

                        requestId = managementRequestService.addManagementRequest(authCertDeletionRequest);
                    } else {
                        var clientRequest = result.getClientRequest()
                                .orElseThrow(() -> new CodedException(X_INVALID_REQUEST, "ClientRequest is missing"));

                        requestId = managementRequestService.addManagementRequest(clientRequest, result.getRequestType());
                    }
                    log.info("Added new management request with id {}", requestId);
                    return requestId;
                });
    }

}
